### PR TITLE
*: support watch from future revision

### PR DESCRIPTION
**Pull request by xiang90**
Fix  https://github.com/coreos/etcd/issues/4624

### CODE REVIEW

File: storage/watchable_store.go [hunk 1]
```diff
@@ -368,9 +368,11 @@ type watcher struct {
...
 	// If end is set, the watcher is on a range.
 	end []byte
-	// cur is the current watcher revision.
-	// If cur is behind the current revision of the KV,
-	// watcher is unsynced and needs to catch up.
+
+	// cur is the current watcher revision of a unsynced watcher.
+	// cur will be updated for unsynced watcher while it is catching up.
+	// cur is startRev of a synced watcher.
+	// cur will not be updated for syunced watcher.
```
Comment (gyuho) [inline storage/watchable_store.go]: typo? `syunced`

File: etcdserver/api/v3rpc/watch.go [hunk 2]
```diff
@@ -119,27 +119,19 @@ func (sws *serverWatchStream) recvLoop() error {
...
 				creq.RangeEnd = []byte{}
 			}
-
-			rev := creq.StartRevision
 			wsrev := sws.watchStream.Rev()
-			futureRev := rev > wsrev
-			if rev == 0 {
-				// rev 0 watches past the current revision
-				rev = wsrev + 1
+			rev := wsrev
```
Comment (heyitsanthony) [inline etcdserver/api/v3rpc/watch.go]: This breaks watchers < currentRev

File: storage/watchable_store_test.go [hunk 3]
```diff
@@ -255,6 +255,39 @@ func TestWatchCompacted(t *testing.T) {
...
+	for i := 0; i < 10; i++ {
+		rev := s.Put(testKey, testValue, lease.NoLease)
+		if rev >= wrev {
+			break
+		}
+	}
+
+	select {
+	case resp := <-w.Chan():
+		if resp.Revision != wrev {
```
Comment (heyitsanthony) [inline storage/watchable_store_test.go]: also check events for correct rev?

File: etcdserver/api/v3rpc/watch.go [hunk 4]
```diff
@@ -119,27 +119,19 @@ func (sws *serverWatchStream) recvLoop() error {
...
 			}
-			// do not allow future watch revision
-			id := storage.WatchID(-1)
-			if !futureRev {
-				id = sws.watchStream.Watch(creq.Key, creq.RangeEnd, rev)
-				if creq.ProgressNotify {
-					sws.progress[id] = true
-				}
+			id := sws.watchStream.Watch(creq.Key, creq.RangeEnd, rev)
+			if creq.ProgressNotify {
```
Comment (heyitsanthony) [inline etcdserver/api/v3rpc/watch.go]: check id != -1?

**Comment by xiang90** [comment]
@heyitsanthony PTAL.

**Comment by heyitsanthony** [comment]
lgtm

**Comment by xiang90** [comment]
@heyitsanthony test failed. but it looks like a grpc issue?

**Pull request merged**